### PR TITLE
[Security] Lazy load guard authenticators and authentication providers

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

Authentication stops on the first authenticator that fails or succeeds, let's instantiate them only if actually needed.

### CODE REVIEW

**Comment by ogizanagi** [comment]
In case a logger is provided, the number of authenticators is provided in the context array ([GuardAuthenticationListener.php#L69](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php#L69)). 
Using `count` on the generator won't provide the right value. I don't think this information is really useful though...

Do you think making `RewindableGenerator` implementing `\Countable` is worth it? (That'll still require an additional check for such cases).

**Comment by nicolas-grekas** [comment]
I think it's a good idea to make `RewindableGenerator` implement Countable, since we always have the info, isn't it?

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by ogizanagi** [comment]
@chalasr , @nicolas-grekas : See #21455

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/GuardAuthenticationFactory.php [hunk 1]
```diff
@@ -15,6 +15,7 @@
 use Symfony\Component\DependencyInjection\ChildDefinition;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\DependencyInjection\Argument\IteratorArgument;
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/GuardAuthenticationFactory.php]: alpha order, should be first I guess

**Review by nicolas-grekas — APPROVED** [review]
:+1: with minor comment

**Comment by chalasr** [comment]
Uses order fixed (must be the last time, moved to a real IDE everywhere)

**Comment by chalasr** [comment]
Same done for authentication providers in last commit https://github.com/symfony/symfony/pull/21450/commits/f3a5447746f70235ca9f86f2a16b84b6f92cfba0. WDYT?

**Comment by chalasr** [comment]
@nicolas-grekas still ok for you?

**Comment by nicolas-grekas** [comment]
Still OK

**Comment by fabpot** [comment]
Thank you @chalasr.

**Pull request merged**